crates/biome_js_analyze/src/lint/nursery/use_explicit_function_return_type.rs

```diff
@@ -266,3 +290,65 @@ fn is_function_used_in_argument_or_expression_list(func: &AnyJsFunction) -> bool
         )
     )
 }
+
+/// Checks whether the given function is a higher-order function, i.e., a function
+/// that returns another function either directly in its body or as an expression.
+///
+/// A higher-order function is one that returns either a regular function or an arrow
+/// function from within its body.
+///
+/// # Arguments
+///
+/// * `func` - A reference to an `AnyJsFunction` that represents the JavaScript function to inspect.
+///
+/// # Returns
+///
+/// * `true` if the function returns another function (either a regular function or an arrow function).
+/// * `false` if it does not return a function or if the body is not a valid returnable function expression.
+///
+/// # Note
+///
+/// This function currently **does not support** detecting a return of a function
+/// inside other statements like `if` statements or `switch` statements. It only detects
+/// direct returns of functions or function returns in a straightforward function body.
+fn is_higher_order_function(func: &AnyJsFunction) -> bool {
+    match func.body().ok() {
+        Some(AnyJsFunctionBody::AnyJsExpression(expr)) => {
+            matches!(
+                expr,
+                AnyJsExpression::JsArrowFunctionExpression(_)
+                    | AnyJsExpression::JsFunctionExpression(_)
+            )
+        }
+        Some(AnyJsFunctionBody::JsFunctionBody(func_body)) => {
+            check_statements_for_function_return(func_body.statements())
+        }
+        _ => false,
+    }
+}
+
+/// Checks whether the given list of JavaScript statements contains a return statement
+/// that returns a function expression (either a regular function or an arrow function).
+///
+/// # Arguments
+///
+/// * `statements` - A list of JavaScript statements (`JsStatementList`) to inspect.
+///
+/// # Returns
+///
+/// * `true` if the list contains a return statement with a function expression as its argument.
+/// * `false` if no such return statement is found or if the list is empty.
+fn check_statements_for_function_return(statements: JsStatementList) -> bool {
+    statements.into_iter().any(|statement| {
+        if let AnyJsStatement::JsReturnStatement(return_stmt) = statement {
+            if let Some(args) = return_stmt.argument() {
+                return matches!(
+                    args,
+                    AnyJsExpression::JsFunctionExpression(_)
+                        | AnyJsExpression::JsArrowFunctionExpression(_)
+                );
+            }
+        }
+        false
+    })
```

Comment:
I think we should check only the first statement because we have to report code like: 

```ts
function f() {
  if (x) {
    return 0;
  }
  return () => {}
}
```

Don't forget to add a test to avoid any regression here.